_posts/2021-03-21-clean-unit-tests-with-mockito.md

```diff
@@ -0,0 +1,375 @@
+* Do not call `reset()` in a test procedure. This is considered as code smell.
+
+## Mockito Specialities and Restrictions
+
+In this section we want to point out important things which are nice to know.
+
+* *What types can I mock?* Mockito allows you to mock not only interfaces but also concrete classes.
+* *What is returned, if I don't declare a mock's behaviour?* Mockito returns `null` for reference objects, and the
+  default values for primitive data types (for example `0` for `int` and `false` for `boolean`)
+* *How many times Mockito returns a previously declared value?* Mockito returns always the same value, regardless of how
+  many times a method is called.
+* *Can I mock `final` classes?* No, final classes **can't** be mocked and neither final methods are mockable. This has
+  to do with the internal mechanism of how Mocktio creates the mock and the Java Language Specification.
+
+### Testing Error Handling with Mockito
+
+```java
+Mockito.when(cityRepository.find(expected.getId())).thenThrow(RuntimeException.class);
+```
+
+Mockito comes with a built-in mechanism to test your error handling. Instead of declaring a return value, advise Mockito
+to throw the expected exception. In case you throw checked exceptions, the compiler doesn't let you to throw checked
+exceptions, that are not declared on the method.
+
+### Mocktio and `void` Methods
+
+```java
+// Causes a compiler error
+Mockito.when(cityRepository.delete()).thenThrow(RuntimeException.class);
+```
+
+In case you want to declare a special behaviour for `void` methods, you must change your approach. The compiler doesn't
+like void methods in brackets, since they have no arguments. Change your approach to following:
+
+```java
+Mockito.doThrow(RuntimeException.class).when(cityRepository).delete(expected);
+```
+
+### Verify Method Calls
+
+```java
+Mockito.verify(cityRepository,Mockito.times(1)).delete(expected);
+```
+
+You can verify how many times a mock was called by simply use the built-in `verify` method. If the condition is not met,
+your test case will fail. This is extremely handy for algorithms or similar processes. There are other predefined
+verification modes such as `atLeastOnce` or `never` already present and ready to use!
+
+## Pros and Cons
+
+Mockito helps you to create simple mocks fast. The application of the methods is easy to read, since they are written in
+fluent style. Mockito can be used in normal projects or in frameworks such as Spring Boot. It Is well documented and has
+lots of examples in it. In case of problems there is a huge community behind and questions are answered frequently on
+StackOverflow. It disposes great flexibility to its users which can contribute their ideas, since it is an open source
+project. Therefore, the development is ongoing, and the project is maintained.
+
+Mockito cant mock constructors, static methods, `equals()` nor `hashCode` out of the box. In order to achieve
```

Comment:
This would be great as a question in the above FAQ section. Here, instead, you can just say that Mockito can't mock everything and there are some restrictions like mocking final and static methods...